### PR TITLE
dbt: only do python setup if the toolchain is freshly unpacked

### PR DESCRIPTION
**Pull request by bobtwinkles**
This has significant performance benefits to DBT startup time (>1 second to <45ms) which is especially helpful when using DBT as a proxy for `clang-format` in an editor's autoformat configuration.

### CODE REVIEW

**Comment by bobtwinkles** [comment]
This needs to be tested by users on Windows, Linux, and OSX before merging

**Comment by m-m-adams** [comment]
Works on mac, time is the same with and without DBT_NO_PYTHON_UPGRADE=1

File: scripts/toolchain/dbtenv.cmd [hunk 1]
```diff
@@ -24,19 +24,24 @@ set "DBT_TOOLCHAIN_ROOT=%DBT_TOOLCHAIN_PATH%\toolchain\win32-x86_64"
 
 set "DBT_TOOLCHAIN_VERSION_FILE=%DBT_TOOLCHAIN_ROOT%\VERSION"
 
+set /p "DBT_TOOLCHAIN_INSTALLED=0"
+
 if not exist "%DBT_TOOLCHAIN_ROOT%" (
     powershell -ExecutionPolicy Bypass -File "%DBT_ROOT%\scripts\toolchain\windows-toolchain-download.ps1" %DBT_TOOLCHAIN_VERSION% "%DBT_TOOLCHAIN_ROOT%"
+    set /p "DBT_TOOLCHAIN_INSTALLED=1"
```
[the remark this answers is not in the document]
Comment (cowboy) [reply scripts/toolchain/dbtenv.cmd:31]: Apparenrtly, `set /p` waits for user input - which makes sense when reading from a file as in `set /p VARNAME=<"%FILENAME%"` but not here!

File: scripts/toolchain/dbtenv.cmd [hunk 2]
```diff
@@ -51,6 +56,25 @@ set "PYTHONNOUSERSITE=1"
...
+        python -c "import wheel"
+        if %ERRORLEVEL% NEQ 0 (
+            set "PIP_WHEEL_PATH=%DBT_TOOLCHAIN_ROOT%\python\wheel"
+            for /R %PIP_WHEEL_PATH% %%G in (
+                certifi*.whl
+            ) do (
+                %PIP_CMD% install -q "%%G"
+            )
+        )
+        %PIP_CMD% install -q --upgrade pip | find /V "already satisfied"
```
Comment (cowboy) [inline scripts/toolchain/dbtenv.cmd:73]: I'm seeing some errors, and I don't know enough to troubleshoot them. Sometimes, the script leaves a `win32-x86_64-10.zip` file and `win32-x86_64` directory in the project root, but I'm not sure how to repro that.

PowerShell:

```
PS C:\Users\Cowboy\projects\DelugeFirmware> .\dbt
Downloading Windows toolchain..done!
Extracting Windows toolchain..moving..done!
Cleaning up temporary files..done!
DBT: Toolchain freshly installed, running python setup
'install' is not recognized as an internal or external command,
operable program or batch file.
```

Git Bash:

```
$ ./dbt.cmd
Downloading Windows toolchain..done!
Extracting Windows toolchain..moving..done!
Cleaning up temporary files..done!
DBT: Toolchain freshly installed, running python setup
install: unknown option -- q
Try 'install --help' for more information.
find: '/V': No such file or directory
find: 'already satisfied': No such file or directory
install: unknown option -- q
Try 'install --help' for more information.
find: '/V': No such file or directory
find: 'already satisfied': No such file or directory
usage: dbt [-h] [<subcommand>]

positional arguments:
  <subcommand>

options:
  -h, --help    print this help message

subcommands:
  Building:
    build [-h] [-S] [-v] [-c] [-t TYPE] [-m] [{release,debug,relwithdebinfo,all}]
      Build the Deluge firmware
    clean [-h]
      Cleanup any build artifacts in the current environment
    configure [-h] [-f] [-m] [-t {dev,nightly,alpha,beta,rc,release}]
      Configure a CMake build (automatically called by build)
    nuke [-h]
      Indiscriminately removes any build tree contents and artifacts
    scons [-h]
      Run SCons in the current environment (deprecated)
  Debugging:
    debug [-h] [-j] [-c COMMAND] [-d TARGET_DEVICE] [-g GDB_PORT] [-v] [-p {swd,jtag}] [-i INTERFACE]
      Run a debug server (JLink or OpenOCD)
    gdb [-h] [-nb] [{7seg,oled}]
      Open a GDB remote for the proper target
    openocd [-h] [-v]
      Run OpenOCD with default arguments (CMSIS-DAP/DelugeProbe).
      These can be changed using the OPENOCD_OPTS variable.
  Development:
    docs [-h] [-R]
      Open the docs for the Deluge Firmware (rebuilding if necessary)
    shell [-h]
      Open a shell in the DBT environment
  Miscellaneous:
    format [-h] [-nr] [-q] [-v] [-c] [directory]
      Formats files using clang-format (Assumes .clang-format present in directory structure)
    license [-h] [-r] [-q] [-d] [-v] directory
      Add Synthstrom Deluge license prelude to files (if not present)
  Ungrouped:
    test [-h]
      Run deluge tests
```

**Review by cowboy — CHANGES_REQUESTED** [review]
Works great here in Linux. It's significantly faster!

In native Windows, however, there are some issue. See my other comments.

File: scripts/toolchain/dbtenv.cmd [hunk 3]
```diff
@@ -24,19 +24,24 @@ set "DBT_TOOLCHAIN_ROOT=%DBT_TOOLCHAIN_PATH%\toolchain\win32-x86_64"
 
 set "DBT_TOOLCHAIN_VERSION_FILE=%DBT_TOOLCHAIN_ROOT%\VERSION"
 
+set /p "DBT_TOOLCHAIN_INSTALLED=0"
```
[the remark this answers is not in the document]
Comment (cowboy) [reply scripts/toolchain/dbtenv.cmd:27]: Apparenrtly, `set /p` waits for user input - which makes sense when reading from a file as in `set /p VARNAME=<"%FILENAME%"` but not here!

File: scripts/toolchain/dbtenv.cmd [hunk 4]
```diff
@@ -24,19 +24,24 @@ set "DBT_TOOLCHAIN_ROOT=%DBT_TOOLCHAIN_PATH%\toolchain\win32-x86_64"
...
+
 if not exist "%DBT_TOOLCHAIN_ROOT%" (
     powershell -ExecutionPolicy Bypass -File "%DBT_ROOT%\scripts\toolchain\windows-toolchain-download.ps1" %DBT_TOOLCHAIN_VERSION% "%DBT_TOOLCHAIN_ROOT%"
+    set /p "DBT_TOOLCHAIN_INSTALLED=1"
 )
 
 if not exist "%DBT_TOOLCHAIN_VERSION_FILE%" (
     powershell -ExecutionPolicy Bypass -File "%DBT_ROOT%\scripts\toolchain\windows-toolchain-download.ps1" %DBT_TOOLCHAIN_VERSION% "%DBT_TOOLCHAIN_ROOT%"
+    set /p "DBT_TOOLCHAIN_INSTALLED=1"
```
[the remark this answers is not in the document]
Comment (cowboy) [reply scripts/toolchain/dbtenv.cmd:36]: Apparenrtly, `set /p` waits for user input - which makes sense when reading from a file as in `set /p VARNAME=<"%FILENAME%"` but not here!

File: scripts/toolchain/dbtenv.cmd [hunk 5]
```diff
@@ -24,19 +24,24 @@ set "DBT_TOOLCHAIN_ROOT=%DBT_TOOLCHAIN_PATH%\toolchain\win32-x86_64"
...
 )
 
 set /p REAL_TOOLCHAIN_VERSION=<"%DBT_TOOLCHAIN_VERSION_FILE%"
 if not "%REAL_TOOLCHAIN_VERSION%" == "%DBT_TOOLCHAIN_VERSION%" (
     echo DBT: starting toolchain upgrade process..
     powershell -ExecutionPolicy Bypass -File "%DBT_ROOT%\scripts\toolchain\windows-toolchain-download.ps1" %DBT_TOOLCHAIN_VERSION% "%DBT_TOOLCHAIN_ROOT%"
     set /p REAL_TOOLCHAIN_VERSION=<"%DBT_TOOLCHAIN_VERSION_FILE%"
+    set /p "DBT_TOOLCHAIN_INSTALLED=1"
```
[the remark this answers is not in the document]
Comment (cowboy) [reply scripts/toolchain/dbtenv.cmd:44]: Apparenrtly, `set /p` waits for user input - which makes sense when reading from a file as in `set /p VARNAME=<"%FILENAME%"` but not here!

**Comment by m-m-adams** [comment]
Works on fedora linux too, so other than windows this should be good to go

**Review by cowboy — APPROVED** [review]
Works great now in PowerShell! I say :shipit:

**Comment by bobtwinkles** [comment]
Cool, thanks everyone for testing. I think this is probably ready to merge?

**Comment by stellar-aria** [comment]
Works in cmd, LGTM
:shipit:

**Review by stellar-aria — APPROVED** [review]

**Pull request merged**